cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/LocalGitServer.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.sourcecontrol;
+
+import org.eclipse.jetty.security.ConstraintMapping;
+import org.eclipse.jetty.security.ConstraintSecurityHandler;
+import org.eclipse.jetty.security.HashLoginService;
+import org.eclipse.jetty.security.UserStore;
+import org.eclipse.jetty.security.authentication.BasicAuthenticator;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.ServerConnector;
+import org.eclipse.jetty.servlet.ServletHandler;
+import org.eclipse.jetty.servlet.ServletHolder;
+import org.eclipse.jetty.util.security.Constraint;
+import org.eclipse.jetty.util.security.Password;
+import org.eclipse.jgit.http.server.GitServlet;
+import org.eclipse.jgit.lib.Repository;
+import org.junit.rules.ExternalResource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nullable;
+import javax.servlet.http.HttpServletRequest;
+
+/**
+ * A local server with authentication that hosts a Git repository for tests.
+ */
+public class LocalGitServer extends ExternalResource {
+  private final Server server;
+  private final GitServlet gitServlet;
+  private final String username;
+  private final String password;
+  private final int port;
+  private static final Logger LOG = LoggerFactory.getLogger(LocalGitServer.class);
+  private ServerConnector serverConnector;
+
+  /**
+   * Creates a {@link LocalGitServer}
+   *
+   * @param username   username for authentication.
+   * @param password   password for authentication.
+   * @param port       port to start the server on. Use 0 for a random port.
+   */
+  public LocalGitServer(String username, String password, int port) {
+    this.server = new Server();
+    this.gitServlet = new GitServlet();
+    this.username = username;
+    this.password = password;
+    this.port = port;
+  }
+
+  @Nullable
+  public String getServerURL() {
+    return serverConnector == null ? null : String.format("http://localhost:%s/", serverConnector.getLocalPort());
+  }
+
+  public void start(Repository repository) throws Exception {
```

Comment:
This should be the `before` method, otherwise it defeated the purpose of using ExternalResource.

Reply:
Done. Didn't do that earlier because the bare git repo initialisation was happening in the test class. I've moved that into the server now.